04-onion-routing.md

```diff
@@ -1407,6 +1438,88 @@ The _origin node_:
   - MAY use the data specified in the various failure types for debugging
   purposes.
 
+## Recommendations for Reputation
+
+Local reputation is used by forwarding nodes to decide whether to endorse a HTLC
+on their outgoing link. Nodes MAY use any metric of their choosing to classify
+a peer as having "good" reputation, though a poor choice of reputation scoring
+metric will affect their downstream reputation.
+
+Peers build reputation by forwarding successful HTLCs that resolve quickly, and
+lose reputation if they endorse failing or slow-resolving HTLCs. Reputation is
+only _negatively_ affected if an endorsed HTLC resolves undesirably, to holds
+nodes accountable for their endorsement signal while still allowing them to 
+forward unendorsed HTLCs that they are not certain about.
+
+HTLC resolution time is assessed relative to a threshold that the node 
+considers to be a reasonable amount of time for a HTLC to resolve: 
+- `resolution_period`: the amount of time a HTLC is allowed to resolve in that 
+  is classified as "good" behavior (default: 10 seconds).
+- `resolution_time`: the time elapsed between `update_add_htlc` and its 
+  resolution (`update_fulfill_hltc` / `update_fail_hltc` / 
+  `update_fail_malformed_htlc`).
+
+Each HTLC's contribution to reputation is expressed by its `effective_fees` 
+which capture the fees that it paid and the opportunity cost that its 
+`resolution_time` incurred. 
+- `fees`: the fees paid by a forwarded HTLC (as described in [BOLT #7](07-routing-gossip.md#htlc-fees), 
+  equal to 0 if the HTLC was failed).
+- `opportunity_cost`: 'ceil ( (resolution_time - resolution_period) / resolution_period) * fees`
+
+Given that `resolution_time` will be > 0 in practice, `opportunity_cost` is 0 
+for HTLCs that resolve within `resolution_period`, and charges the `fees` that 
+the HTLC would have earned per period it is held thereafter. This cost accounts 
+for the slot and liquidity that could have otherwise been paid for by 
+successful, fast resolving HTLCs during the `resolution_time` the HTLC was 
+locked in the channel.
+
+For every incoming HLTC a peer has forwarded through a node, its `effective_fees` 
+are calculated as follows:
+- if `endorsed` = 1 in `update_add_htlc`:
+  - `effective_fees` = `fees` - `opportunity_cost` 
```

Comment:
Perhaps this could be a bit more rewarding by e.g applying a multiplier to `fees`?

I wonder if this slow to gain but fast to lose reputation will lead to a state in the network where the majority of HTLCs are not being endorsed (at least not e2e). You can always rely on the successful `unkown` HTLCs to eventually build up some good relations between nodes but the punishment could very fast deplete the reputation on those short-lived endorsed links.

It's tough to pick the right values for `routing_window` & `resolution_period`, especially because they need to be fine tuned according to the network activity which is very hard to figure out (we don't really know what the network-wide average HTLC resolution time is for example). Plus it will be hard to gather reputation related stats after the upgrade is out so perhaps these values could be dynamic?

Reply:
> Perhaps this could be a bit more rewarding by e.g applying a multiplier to fees?

I'd be cautious to add any type of multiplier here, we don't want to allow attackers to build up disproportionate "credit" with a brief period of good behaviour and then abuse it.

> I wonder if this slow to gain but fast to lose reputation will lead to a state in the network where the majority of HTLCs are not being endorsed (at least not e2e)

I don't think this is terrible? Nodes that want to "play it safe" can forward `endorsed=0` traffic (at the cost of having access to fewer resources), and those more committed to managing their node + traffic will set `endorsed=1` where appropriate?

> It's tough to pick the right values for routing_window

Indeed, well managed nodes will need to chose how they set these params per their activity / risk preference - no way we're going to be able to make a wholesale guess that works for everyone. 

> It's tough to pick the right values for ... resolution_period

For `resolution_period` I think we can just play it safe and pick a default that is reasonable in most circumstances (eg: network diameter `X`, worst case per-hop resolution time `Y`, `resolution_period = X*Y`), even if we overestimate the default we're still penalizing the more agregious slow jamming attacks that would hold for way longer.

Reply:
> I'd be cautious to add any type of multiplier here, we don't want to allow attackers to build up disproportionate "credit" with a brief period of good behavior and then abuse it.

Agree. Maybe worth noting that this proposal makes network-wide jam attack nearly impossible, but still leaves space for targeted channel jams, and this is where `fee` multiplier could make things easier for the attacker. :+1: 

> I don't think this is terrible? Nodes that want to "play it safe" can forward endorsed=0 traffic (at the cost of having access to fewer resources), and those more committed to managing their node + traffic will set endorsed=1 where appropriate?

It's not terrible, I'm just wondering if the current punishment/reward system is going to "trap" nodes in a neutral state. Theoretically that could lock up 50% of the network's throughput. Plus, what we want to ideally achieve is the majority of payments to still be able to go through the network regardless of if there's an ongoing network jamming attack. If the majority of "honest" nodes have a hard time acquiring the `endorsed` flag from their peers then they'll have to fall back to the `unknown_allocation_slots` which is where the attacker is also taking action.

Reply:
>  I'm just wondering if the current punishment/reward system is going to "trap" nodes in a neutral state

WDYT about `effective_fees` for successful but `endorsed=0` HTLCs only being some fraction of fees (handwave, 50%)? That came up as an option when we were discussing this. Would make it slower to build up reputation without endorsing, but adds some incentive to endorse payments.

Reply:
> WDYT about effective_fees for successful but endorsed=0 HTLCs only being some fraction of fees (handwave, 50%)?

Sgtm, we need to incentivise using `endorsed=1` in low traffic environments (when peer is "good"). 50% looks good :shrug: 

> That came up as an option when we were discussing this. Would make it slower to build up reputation without endorsing, but adds some incentive to endorse payments.

Yes I believe this doesn't heavily damage the passive build-up of reputation in a "neutral" network, there is still some passive push towards using 100% of the network's bandwidth.

Reply:
Pushed a fixup adding discount as suggested!